CenterForOpenScience/osf.io
# Feature/Refactor Citations for APIV2 [PLAT-450]

## Purpose

Move folder listing functionality, currently (typically) located in addons/<addon>/views.py to the NodeSettings model to make it accessible to APIv2. This will likely require refactoring the generic_views, and tests as well.  This ticket is for Mendeley and Zotero.

## Changes

- For Mendeley, show top-level folders only for now.  [ From Mendeley API docs:](http://dev.mendeley.com/code/core_quick_start_guides.html):

> It is not possible to list the child folders of a folder directly using the API. Your application should construct a tree model of folders by parsing the results of /folders

Currently returning all folders in a flat manner.  The path lists the parent folder id.
`http://localhost:8000/v2/nodes/tf2mq/addons/mendeley/folders/`
<img width="619" alt="screen shot 2018-04-03 at 5 43 00 pm" src="https://user-images.githubusercontent.com/9755598/38279571-7e730612-3766-11e8-9b23-60fbce3458c8.png">


- For Zotero, show all levels of folders. Top-level is available libraries, which is your personal library plus group libraries. Show secondary tiers which are folders within particular libraries, and then folders within folders.  The path is the top-level library_id.

### Top level - Libraries
<img width="792" alt="screen shot 2018-04-03 at 4 50 36 pm" src="https://user-images.githubusercontent.com/9755598/38277843-78d25e4e-375f-11e8-96dc-2ca1999ee1b3.png">

### Secondary level - top-level folders within a library
<img width="744" alt="screen shot 2018-04-03 at 4 50 50 pm" src="https://user-images.githubusercontent.com/9755598/38277846-7c18c606-375f-11e8-969a-26f5d271ee94.png">

### Third level - subfolders within a folder
<img width="820" alt="screen shot 2018-04-03 at 4 51 14 pm" src="https://user-images.githubusercontent.com/9755598/38277850-7f2638ec-375f-11e8-9203-7c2d364848a8.png">

### And so on... when there are no more folders at a sublevel, returns an empty list.


## QA Notes

For Mendeley, create a couple of folders. For Zotero, create a couple of libraries, and folders within libraries.  Check API endpoint `http://localhost:8000/v2/nodes/node_id/addons/mendeley/folders/` and `http://localhost:8000/v2/nodes/node_id/addons/zotero/folders/`.  Mendeley will return all folders flat for now, and have children link removed.  Zotero will have children links.  If you follow the children links you will see all the nested folders.

Since Mendeley and Zotero API v1 views have been refactored to use these same methods as v2, I would do a full regression on both Mendeley and Zotero to make sure nothing has changed.

## Side Effects



## Ticket
https://openscience.atlassian.net/browse/PLAT-450
